internal/core/workers/metricsreporter/metrics_reporter_worker.go

```diff
@@ -0,0 +1,163 @@
+		logger:                        zap.L().With(zap.String(logs.LogFieldServiceName, "metrics_reporter_worker"), zap.String(logs.LogFieldSchedulerName, scheduler.Name)),
+	}
+}
+
+// Start is responsible for starting a loop that will
+// periodically report metrics for scheduler pods and game rooms.
+func (w *MetricsReporterWorker) Start(ctx context.Context) error {
+	w.workerContext, w.cancelWorkerContext = context.WithCancel(ctx)
+	ticker := time.NewTicker(time.Millisecond * w.metricsReporterIntervalMillis)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-w.workerContext.Done():
+			w.Stop(w.workerContext)
+			return nil
+		case <-ticker.C:
+			w.reportInstanceMetrics()
+			w.reportGameRoomMetrics()
+		}
+	}
+}
+
+func (w *MetricsReporterWorker) Stop(_ context.Context) {
+	if w.workerContext == nil {
+		return
+	}
+
+	w.cancelWorkerContext()
+}
+
+func (w *MetricsReporterWorker) IsRunning() bool {
+	return w.workerContext != nil && w.workerContext.Err() == nil
+}
+
+func (w *MetricsReporterWorker) reportInstanceMetrics() {
+	w.logger.Info("Reporting instance metrics")
+
+	instances, err := w.instanceStorage.GetAllInstances(w.workerContext, w.schedulerName)
+	if err != nil {
+		w.logger.Error("Error getting pods", zap.Error(err))
+		return
+	}
+	readyInstances, pendingInstances, errorInstances, unknownInstances, terminatingInstances := 0, 0, 0, 0, 0
+	for _, instance := range instances {
+		switch instance.Status.Type {
+		case game_room.InstanceReady:
+			readyInstances++
+		case game_room.InstancePending:
+			pendingInstances++
+		case game_room.InstanceError:
+			errorInstances++
+		case game_room.InstanceUnknown:
+			unknownInstances++
+		case game_room.InstanceTerminating:
+			terminatingInstances++
+		}
+	}
+	reportInstanceReadyNumber(w.schedulerName, readyInstances)
+	reportInstancePendingNumber(w.schedulerName, pendingInstances)
+	reportInstanceErrorNumber(w.schedulerName, errorInstances)
+	reportInstanceUnknownNumber(w.schedulerName, unknownInstances)
+	reportInstanceTerminatingNumber(w.schedulerName, terminatingInstances)
+
+}
+
+func (w *MetricsReporterWorker) reportGameRoomMetrics() {
+	w.logger.Info("Reporting game room metrics")
+
+	readyRooms, err := w.roomStorage.GetRoomCountByStatus(w.workerContext, w.schedulerName, game_room.GameStatusReady)
+	if err != nil {
+		w.logger.Error("Error getting ready pods", zap.Error(err))
+	} else {
+		reportGameRoomReadyNumber(w.schedulerName, readyRooms)
+	}
+	pendingRooms, err := w.roomStorage.GetRoomCountByStatus(w.workerContext, w.schedulerName, game_room.GameStatusPending)
+	if err != nil {
+		w.logger.Error("Error getting pending pods", zap.Error(err))
+	} else {
+		reportGameRoomPendingNumber(w.schedulerName, pendingRooms)
+	}
+	occupiedRooms, err := w.roomStorage.GetRoomCountByStatus(w.workerContext, w.schedulerName, game_room.GameStatusOccupied)
+	if err != nil {
+		w.logger.Error("Error getting occupied pods", zap.Error(err))
+	} else {
```

Comment:
What is your opinion about implementing tiny functions instead of going to the if-else list approach? I think it makes it harder to read/debug a big function with very similar conditions. Also, since it happens in the else condition, my first reaction was to think that it's not the expected behavior
Like:
```
func (w *MetricsReporterWorker) reportGameRoomMetrics() {
        w.reportReadyRooms()
        w.reportPendingRooms()
        w.reportTerminatingRooms()
        ...
}

func (w *MetricsReporterWorker) reportReadyRooms () {
        readyRooms, err := w.roomStorage.GetRoomCountByStatus(w.workerContext, w.schedulerName, game_room.GameStatusReady)
	if err != nil {
		w.logger.Error("Error getting ready pods", zap.Error(err))
		return
	}
	reportGameRoomReadyNumber(w.schedulerName, readyRooms)
}
...
```

Reply:
I like this approach!